File: DekafMain/src/text/TextFragment.java
package org.jetbrains.dekaf.text;

import org.jetbrains.annotations.NotNull;

import java.io.Serializable;



/**
 * @author Leonid Bushuev from JetBrains
 **/
public class TextFragment implements Comparable<TextFragment>, Serializable {

  @NotNull
  public final String text;

  /**
   * Row number, starts with 1.
   */
  public final int row;

  /**
   * Number of the position inside the row, starts with 1.
   */
  public final int pos;


  /**
   * Instantiates a new TextFragment.
   * @param text  text fragment itself.
   * @param row   the row number, starts with 1.
   * @param pos   the number of the position inside the row, starts with 1.
   */
  public TextFragment(@NotNull final String text, final int row, final int pos) {
    this.text = text;
    this.row = row;
    this.pos = pos;
  }


  @NotNull
  public String getTextName() {
    return "SQL";
  }


  @Override
  public int compareTo(@NotNull final TextFragment that) {
    if (this == that) return 0;
    //noinspection ConstantConditions
    if (that == null) throw new IllegalArgumentException("Comparison with null");

    int z =         this.getTextName().compareTo(that.getTextName());
    if (z == 0) z = this.row - that.row;
    if (z == 0) z = this.pos - that.pos;
    if (z == 0) z = this.text.compareTo(that.text);
    return z;
  }


  @Override
  public boolean equals(final Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    TextFragment that = (TextFragment) o;

    return this.row == that.row
        && this.pos == that.pos
        && this.text.equals(that.text);
  }

  @Override
  public int hashCode() {
    return text.hashCode();
  }


  @NotNull
  protected String getTextLocationHumanReadable() {
    return "Fragment at " + row + ':' + pos;
  }


  @Override
  public String toString() {
    return getTextLocationHumanReadable() + '\n'
        + text;
  }

}

File: dekaf-main/src/text/TextFragment.kt
package org.jetbrains.dekaf.text

import java.io.Serializable


/**
 * Fragment of a text.
 */
open class TextFragment
(
        /**
         * The text of this fragment.
         */
        val text: String,

        /**
         * Row number, starts with 1.
         */
        val row: Int,

        /**
         * Number of the position inside the row, starts with 1.
         */
        val pos: Int
)
    : Comparable<TextFragment>, Serializable
{


    open val textName: String
        get() = "SQL"


    override fun compareTo(other: TextFragment): Int {
        if (this === other) return 0

        var z = this.textName.compareTo(other.textName)
        if (z == 0) z = this.row - other.row
        if (z == 0) z = this.pos - other.pos
        if (z == 0) z = this.text.compareTo(other.text)
        return z
    }


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val that = other as TextFragment

        return this.row == that.row
            && this.pos == that.pos
            && this.text == that.text
    }

    override fun hashCode(): Int {
        return text.hashCode()
    }


    protected open val textLocationHumanReadable: String
        get() = "Fragment at $row:$pos"


    override fun toString(): String {
        return textLocationHumanReadable + '\n' + text
    }

}
